### DISCUSSION TITLE
How to use real sensor values from ESP in Berry code

### DISCUSSION THREAD

**Discussion by Bournetoride** [arendst/Tasmota#17911] [Q&A] [answered]
I need to use multiple DS18B20 temperatures in my Berry code calculations. These are local sensors, not from MQTT messages.
Tried using berry's tasmota.add_rule to capture the temperature events  with
def rule_dsb(value, trigger)
var i=string.find(trigger,"DS18B20-")
var tr=string.split(trigger,i+46) # DS18B20-1":{"Id":"0218917752DC","Temperature":17.9
var dsnum=number.split(trigger,i+8)
var temp=number(tr[4])
var DS18B20_(dsnum) = temp
end

tasmota.add_rule("DS18B20-1",rule_dsb)
tasmota.add_rule("DS18B20-2",rule_dsb)
tasmota.add_rule("DS18B20-3",rule_dsb)

but console says
00:00:00.413 BRY: failed to run compiled code 'syntax_error' - /heating.be:399: unexpected token 'tasmota'

**Comment by sfromis** [comment]
Since your error message says line 399, look at that line in context, not visible above. And I don't think that you need `import tasmota`.

Anyway, I'd recommend you to not use that string splitting technique, but use the already parsed JSON into a map data structure. Much nicer that way.

Like this:
```
tasmota.add_rule('DS18B20-1',
  def (value,trigger,data)
    print('Temperatures', data['DS18B20-1']['Temperature'], data['DS18B20-2']['Temperature'], data['DS18B20-3']['Temperature'])
  end
)
```

**Reply by barbudor** [reply]
Indeed, my bad, 'import tasmota' not needed

**Comment by Bournetoride** [comment, marked as answer]
Yes  import tasmota is not required. I tried tasmota.add_rule('DS18B20-1'..... and the error disappears
Does this give me variables I can use to calculate the required setpoint, it will be (DS18B20-1 +5'C from sunrise until 13:00 then shift gradually to DS18B20-2 +10'C at sunset) when DS18B20-3 exceeds setpoint then turn on relay
Hysteresis to be sorted later

**Reply by sfromis** [reply]
Not sure what you specifically ask about. The example I posted shows that the single add_rule function could get temperature values from 3 sensors, without messing around with chopping up strings.

**Reply by Bournetoride** [reply]
I need to have variables I can use in berry code calculations, does "print('Temperatures', data['DS18B20-1']['Temperature'], data['DS18B20-2']['Temperature'], data['DS18B20-3']['Temperature'])" do that, or does it just print them in a console, I'm struggling to understand some Berry syntax, hoping to learn by using it

**Reply by sfromis** [reply]
That specific print is just an example to demonstrate availability. 
You could also do normal Berry things like `var temperature1 = data['DS18B20-1']['Temperature']`

**Reply by Bournetoride** [reply]
var temperature1 = data['DS18B20-1']['Temperature'] 
thats the answer, so simple, thanks
